src/transformers/generation_beam_constraints.py

```diff
@@ -202,6 +201,145 @@ def copy(self, stateful=False):
         return new_constraint
 
 
+class DisjunctiveTrie:
+    def __init__(self, nested_token_ids: List[List[int]], no_subsets=True):
+        r"""
+        A helper class that builds a trie with the words represented in `nested_token_ids`.
+
+        For example, if `nested_token_ids==[[1,2,3], [1,2,4], [1,5,6]]`, then the trie is: 1 -> 2 -> 3
+                            \ -> 4 -> 5 -> 6
+        """
+        self.max_height = max([len(one) for one in nested_token_ids])
+
+        leaves = []
+        root = dict()
+        for token_ids in nested_token_ids:
+            level = root
+            for tidx, token_id in enumerate(token_ids):
+                if id(level) in leaves:
+                    raise ValueError(
+                        f"Each list in `nested_token_ids` can't be a complete subset of another list, but is {nested_token_ids}."
+                    )
+
+                if token_id not in level:
+                    level[token_id] = dict()
+
+                level = level[token_id]
+
+                if tidx == len(token_ids) - 1:
+                    leaves.append(id(level))
```

Comment:
Right, that was how I went about it the first time. But I guess I just found it inefficient because it would necessarily have to traverse the full graph again. I agree with all your comments. I'll make a final edit on this part.